python/nammu/view/AtfAreaView.py

```diff
@@ -274,15 +274,49 @@ def keyTyped(self, ke):
 
 class AtfUndoableEditListener(UndoableEditListener):
     '''
-    Overrides the undoableEditHappened functionality to only count INSERT type
-    events instead of capturing other changes and events like highlighting, etc.
+    Overrides the undoableEditHappened functionality to group INSERT/REMOVE 
+    edit events with their associated CHANGE events (these correspond to 
+    highlighting only at the moment).
+    TODO: Make compounds save whole words so undoing is not so much of a pain 
+          for the user.
     '''
     def __init__(self, undo_manager):
         self.undo_manager = undo_manager
+        self.current_compound = CompoundEdit()
+        self.must_compound = False
+
+
+    def compound(self):
+        '''
+        Allows for external actions like setText() which adds two consecutive
+        INSERT events to manually start and stop a compound from outside
+        the undoableEditHappened() call.
+        '''
+        self.must_compound = True
+        self.current_compound.end()
+        self.undo_manager.addEdit(self.current_compound)
+        self.current_compound = CompoundEdit()
+
+        
+    def stop_compound(self):
+        self.must_compound = False
+        self.current_compound.end()
+        self.undo_manager.addEdit(self.current_compound)       
 
     def undoableEditHappened(self, event):
         edit = event.getEdit()
-        edit_type = edit.getType()
-
-        if (str(edit_type) == "INSERT" or str(edit_type) == "REMOVE") and edit.significant:
-            self.undo_manager.addEdit(edit)
+        edit_type = str(edit.getType())
+
+        # If significant INSERT/REMOVE event happen, end and add current
+        # edit compound to undo_manager and start a new one.
+        if (edit_type == "INSERT" or edit_type == "REMOVE") \
+            and not self.must_compound:
+            # Explicitly end compound edits so their inProgress flag goes to
+            # false. Note undo() only undoes compound edits when they are not
+            # in progress.
+            self.current_compound.end()
+            self.current_compound = CompoundEdit()
+            self.undo_manager.addEdit(self.current_compound)
+            
+        # Always add current edit to current compound  
```

Comment:
I think a context manger like the following should work but it's completly untested



```python
import contextlib

@contextlib.contextmanager
def force_compound(self):
    self.must_compound = False
    self.current_compound.end()
    self.undo_manager.addEdit(self.current_compound)  
    try:
        yield
    finally:
        self.must_compound = False
        self.current_compound.end()
        self.undo_manager.addEdit(self.current_compound)    
```